### PR TITLE
Prefer `e.is_number` over `not e.free_symbols` in SymPy

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #112688

We spend somewhere on the order 1% in `sympy.Expr.free_symbols` as it is called millions of times.
Most of the time we actually just want to know "is this a constant", however `e.is_constant()` is
horribly slow. It turns out though that there is another propery `is_number` that does what we want.

> property is_number:
>
> Returns True if self has no free symbols and no undefined functions (AppliedUndef, to be precise). It will be faster
> than if not self.free_symbols, however, since is_number will fail as soon as it hits a free symbol or undefined
> function.

Even further, we also avoid the overhead of building the unnecessary set object.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: torch/fx/experimental/symbolic_shapes.py [hunk 1]
```diff
@@ -151,26 +152,36 @@ def is_concrete_bool(a: Union[bool, SymBool]):
...
+    # we need at least 1 to call union, so we hand code the identity
+    try:
+        first_expr = next(itr)
+    except StopIteration:
+        return set()
+
+    return first_expr.free_symbols.union(*(e.free_symbols for e in itr))
+
+def has_free_symbols(val: Union[SymInt, torch.Tensor]) -> bool:
+    """Equivalent to bool(free_symbols(val))"""
```
Comment (lezcano) [inline torch/fx/experimental/symbolic_shapes.py:183]: perhaps
```suggestion
    """Faster version of bool(free_symbols(val))"""
```

File: torch/_inductor/utils.py [hunk 2]
```diff
@@ -1215,8 +1215,8 @@ def is_linux() -> bool:
     return platform.system() == "Linux"
 
 
-def has_free_symbols(itr):
-    return any(hasattr(x, "free_symbols") and len(x.free_symbols) > 0 for x in itr)
+def has_free_symbols(itr: Iterable[Any]):
+    return any(isinstance(x, sympy.Expr) and not x.is_number for x in itr)
```
Comment (lezcano) [inline torch/_inductor/utils.py:1219]: A non-equivalent definition? Perhaps better to change the name of this one or try to remove it?

Comment (lezcano) [reply torch/_inductor/utils.py:1219]: It's not equivalent in general, but I think that they may be equivalent in their callsites.

**Review by lezcano — APPROVED** [review]

File: torch/_inductor/utils.py (hunk 2 above)
Comment (peterbell10) [reply torch/_inductor/utils.py:1219]: I think it's mainly used by the `is_dynamic` function defined below.  In which case it's close enough in space to eliminate confusion.

**Pull request closed without merging**